src/app/shared-services/inferences/engine/staypoint-engine/staypoint-engine.ts

```diff
@@ -127,23 +127,142 @@ export class StaypointEngine implements IInferenceEngine {
     return inferences
   }
 
+  /**
+   * @param trajectory a user trajectory
+   * @returns the number of days during which the trajectory was recorded
+   */
   private countDays(trajectory: Trajectory): number {
-    // TODO update this method when start/stop are encoded in trajectory
-    const diffMs =
-      trajectory.timestamps[trajectory.timestamps.length - 1].getTime() -
-      trajectory.timestamps[0].getTime()
+    if (!trajectory.state) {
+      return this.countContinuousDays(
+        trajectory.timestamps[0],
+        trajectory.timestamps[trajectory.timestamps.length - 1]
+      )
+    }
+    // find indices where trajectory has started
+    const startIndices = []
+    let idx = 0
+    while (idx !== -1) {
+      startIndices.push(idx)
+      idx = trajectory.state.indexOf(PointState.START, idx + 1)
+    }
+    // process each section after start individually
+    let sumDays = 0
+    for (let i = 0; i++; i < startIndices.length - 1) {
+      const date1 = trajectory.timestamps[startIndices[i]]
+      const date2 = trajectory.timestamps[startIndices[i + 1] - 1]
+      sumDays += this.countContinuousDays(date1, date2)
+      // countContinuousDays counts every started day as one, so we need to
+      // make sure we dont double count for start/end in same day
+      if (i !== 0) {
+        const prevDate = trajectory.timestamps[startIndices[i] - 1]
+        if (
+          date1.getDate() === prevDate.getDate() &&
+          date1.getMonth() === prevDate.getMonth() &&
+          date1.getFullYear() === prevDate.getFullYear()
```

Comment:
These 3 lines are seemingly used a couple of times, so: could it be useful to create a separate method for this?